ecl/hql/hqlexpr.cpp

```diff
@@ -5400,113 +5375,59 @@ bool CUsedTables::usesSelector(IHqlExpression * selector) const
 
 void CUsedTables::gatherTablesUsed(CUsedTablesBuilder & used) const
 {
-    if (numTables == 0)
+    if (numActiveTables == 0)
         return;
-    if (numTables == 1)
+    if (numActiveTables == 1)
     {
-        if (numActiveTables == 1)
-            used.addActiveTable(tables.single);
-        else
-            used.addNewTable(tables.single);
+        used.addActiveTable(tables.single);
     }
     else
     {
         for (unsigned i1=0; i1 < numActiveTables; i1++)
             used.addActiveTable(tables.multi[i1]);
-        for (unsigned i2=numActiveTables; i2 < numTables; i2++)
-            used.addNewTable(tables.multi[i2]);
     }
 }
 
-void CUsedTables::gatherTablesUsed(HqlExprCopyArray * newScope, HqlExprCopyArray * inScope) const
+void CUsedTables::gatherTablesUsed(HqlExprCopyArray & inScope) const
 {
-    if (numTables == 0)
+    if (numActiveTables == 0)
         return;
-    if (numTables == 1)
+    if (numActiveTables == 1)
     {
-        if (numActiveTables == 1)
-        {
-            if (inScope)
-                addUniqueTable(*inScope, tables.single);
-        }
-        else
-        {
-            if (newScope)
-                addUniqueTable(*newScope, tables.single);
-        }
+        addUniqueTable(inScope, tables.single);
     }
     else
     {
-        if (inScope)
-        {
-            for (unsigned i1=0; i1 < numActiveTables; i1++)
-                addUniqueTable(*inScope, tables.multi[i1]);
-        }
-        if (newScope)
-        {
-            for (unsigned i2=numActiveTables; i2 < numTables; i2++)
-                addUniqueTable(*newScope, tables.multi[i2]);
-        }
+        for (unsigned i1=0; i1 < numActiveTables; i1++)
+            addUniqueTable(inScope, tables.multi[i1]);
     }
 }
 
 
-void CUsedTables::set(HqlExprCopyArray & activeTables, HqlExprCopyArray & newTables)
+void CUsedTables::set(HqlExprCopyArray & activeTables)
 {
     numActiveTables = activeTables.ordinality();
-    numTables = numActiveTables + newTables.ordinality();
-    if (numTables == 1)
+    if (numActiveTables == 1)
     {
-        if (numActiveTables == 1)
-        {
-            tables.single = &activeTables.item(0);
-        }
-        else
-        {
-            tables.single = &newTables.item(0);
-            tables.single->Link();
-        }
+        tables.single = &activeTables.item(0);
     }
-    else if (numTables != 0)
+    else if (numActiveTables != 0)
     {
-        IHqlExpression * * multi = new IHqlExpression * [numTables];
+        IHqlExpression * * multi = new IHqlExpression * [numActiveTables];
```

Comment:
@ghalliday Where is this (new IHqlExpression * [numActiveTables]) released?

Reply:
well spotted - I misread your original comment.  I will fix.